apm-agent-attach/src/main/java/co/elastic/apm/attach/ElasticApmAttacher.java

```diff
@@ -74,14 +75,23 @@ public static void attach() {
     /**
      * Attaches the Elastic Apm agent to the current JVM.
      * <p>
-     * This method may only be invoked once.
+     * This method should only be invoked once.
+     * Subsequent calls will be ignored.
      * </p>
      *
      * @param configuration the agent configuration
      * @throws IllegalStateException if there was a problem while attaching the agent to this VM
      */
     public static void attach(Map<String, String> configuration) {
-        ByteBuddyAgent.attach(AgentJarFileHolder.INSTANCE.agentJarFile, ByteBuddyAgent.ProcessProvider.ForCurrentVm.INSTANCE, toAgentArgs(configuration));
+        synchronized (ElasticApmAttacher.class) {
```

Comment:
It's not really thread safe because this system property is set elsewhere.
We can kind of assume that this wouldn't be called again before `co.elastic.apm.agent.bci.AgentMain#init` is invoked, but if that's the case- no need to make it `synchronized`.

I suggest just in order to stay on the safe side and avoid over-thinking this, you can just use a different property here and there...

Reply:
It's set in a method which is `synchronized` as well. I don't think there's a need for `attach` to be called concurrently. It seems safer to enforce this to be handled by one thread at a given time. I'd also like to avoid have two system properties for essentially the same use-case.

Reply:
This block is synchronized on `ElasticApmAttacher.class` and the other is synchronzed on `AgentMain.class`. Maybe just sync them on the same monitor?

Reply:
I don't think there is a need to do that. There's also no cyclic dependency so that there's no danger of deadlocks.